### PR TITLE
Make nicer pull request merge messages

### PR DESCRIPTION
**Pull request by btcdrak**
Produces merge commit messages as follows

```
Merge #12345: Pull request title
```

I've written this code semi blind because I'm not sure how to test it in the final run, but essentially the magic comes from taking the title from Github API. 

Needs testing.

### CODE REVIEW

**Comment by jonasschnelli** [comment]
Nice!

utACK

**Comment by btcdrak** [comment]
Can someone restart the travis build please https://travis-ci.org/bitcoin/bitcoin/jobs/46357658 - the job timed out.

**Comment by sipa** [comment]
Can you use the env variables for user and repository name? This script is
not solely used for Bitcoin Core.

**Comment by btcdrak** [comment]
@sipa Sure. I didnt see a `$USER` variable though, only `$REPO`.

**Comment by btcdrak** [comment]
Scrub that, I see that `$REPO` is `user/repo`

**Comment by btcdrak** [comment]
Done.

**Comment by btcdrak** [comment]
@sipa can we assume this script is always for github projects? Otherwise we'd need something like this https://gist.github.com/btcdrak/bd20efaec5ac3e5e6a92

**Comment by laanwj** [comment]
@btcdrak it's called github-merge.sh ;)

Concept ACK, haven't tested.

File: contrib/devtools/github-merge.sh [hunk 1]
```diff
@@ -83,12 +83,13 @@ function cleanup() {
 
 # Create unsigned merge commit.
 (
-  echo "Merge pull request #$PULL"
+  PRTITLE=`curl -s https://api.github.com/repos/$REPO/pulls/$PULL | grep -e '  "title": ".*",'| awk -F'"' '{print $4}'`
```
Comment (fanquake) [inline contrib/devtools/github-merge.sh]: This needs to have the organisation in the URL. i.e api.github.com/repos/bitcoin/bitcoin/pulls/5623 As is, it just returns "Not Found" messages.

Comment (btcdrak) [inline contrib/devtools/github-merge.sh]: $REPO is supposed to be something like `bitcoin/bitcoin` as per [this](https://github.com/btcdrak/bitcoin/blob/master/contrib/devtools/github-merge.sh#L15-L18)

Comment (fanquake) [inline contrib/devtools/github-merge.sh]: Right, forgot we already had that defined. Although I still can't get the script to run. Only get "ERROR: Creating merge failed (already merged?)."

Comment (btcdrak) [inline contrib/devtools/github-merge.sh]: Works for me https://github.com/btcdrak/bitcointest/pull/1

```
$:~/bitcoin/contrib/devtools$ git config --global user.signingkey EA8451F1
$:~/bitcoin/contrib/devtools$ ./github-merge.sh 1
Dropping you on a shell so you can try building/testing the merged source.
Run 'git diff HEAD~' to show the changes being merged.
Type 'exit' when done.
(1)$:~/bitcoin/contrib/devtools$ exit
exit
Press 'm' to accept the merge. m
Merge accepted.
Press 's' to sign off on the merge. s
```

Comment (btcdrak) [inline contrib/devtools/github-merge.sh]: Disregard last message, there is a problem.

Comment (btcdrak) [inline contrib/devtools/github-merge.sh]: Fixed. TIL: bash has scoping.

**Comment by btcdrak** [comment]
Example tested here: https://github.com/btcdrak/bitcointest/commit/87582f4e35997e7d6beef6578438a0dab252debe from btcdrak/bitcointest#2

File: contrib/devtools/github-merge.sh [hunk 2]
```diff
@@ -82,13 +82,15 @@ function cleanup() {
 }
 
 # Create unsigned merge commit.
+PRTITLE=`curl -s https://api.github.com/repos/$REPO/pulls/$PULL | grep -e '  "title": ".*",'| awk -F'"' '{print $4}'`
```
Comment (laanwj) [inline contrib/devtools/github-merge.sh]: I think this will fail with "'s in the pull title?

Comment (btcdrak) [inline contrib/devtools/github-merge.sh]: @laanwj It works for me. I posted the test merge examples above.

Comment (btcdrak) [inline contrib/devtools/github-merge.sh]: Another example 
PR: https://github.com/btcdrak/bitcointest/pull/3
Merge commit: https://github.com/btcdrak/bitcointest/commit/fb41e01abcca2ca0e52ed64e7cd04dd256cf7ad4

Comment (laanwj) [inline contrib/devtools/github-merge.sh]: But none of those titles has a " (or quote) symbol in it, or am I missing something?

**Comment by laanwj** [comment]
E.g. I've tried merging https://github.com/btcdrak/bitcointest/pull/4 using your script, which results in a chopped-off title:

    Merge #4: \

I'm kind of distrustful of these kinds of constructions in bash script. It's very hard to make them cope with all given input without introducing bugs. As this is data coming from the network this problem is compounded, and may even result in vulnerabilities.

**Comment by btcdrak** [comment]
@laanwj Bash certainly isn't ideal here, and maybe I should use a helper python script to fetch the data, check it's valid json and can then extract the correct field. The current solution is just trying to identify the following pattern from the API call rather than parse the json.

    "title": "pr title goes here",

**Comment by laanwj** [comment]
Another idea in this direction (but even hackier to implement in shellscript - maybe we should bite the bullet and rewrite this script in Python some day) would be to include the pull request message in the body of the pull merge message. For larger projects, developers tend to write 'higher level' context information in the pull request message, that's not in the commits themselves.

Not arguing that you should do this here, better is better.

**Comment by laanwj** [comment]
Had to revert this (aaba10f) due to various issues
    
- Pull request names get cut off at ", see e.g. a026a56
- Merge script no longer copes with pulls that have a milestone
      attached, due to a duplicate 'title' in JSON that is not handled by the
      ad-hoc parsing.

If anyone is still interested in doing this, I'd suggest rewriting the script in a language that can do real JSON parsing, preferably Python.

**Comment by sipa** [comment]
For posteriority: this was improved in #7378

**Pull request merged**